### PR TITLE
bug: corr CZ-41 - Fix bank account unique naming

### PR DESCRIPTION
**Pull request by silviczka**
- Fix bank account unique naming so duplicate names get the correct next suffix and single-digit suffixes like AKA (1) are treated as user text, not system-generated suffixes.
- Enforce active per-user bank account name uniqueness at the database/write path with retry handling for concurrent create and rename collisions.
- Add unit test coverage for the naming rules, including per-user scoping and display-name splitting edge cases.
- Extract pure bank-account name display logic into a shared helper so UI components no longer depend on the domain module for formatting concerns.
- Update the UI to display system-generated suffixes in muted styling and show the actual final bank account name in success toasts, including when the requested name had to be adjusted.

## Summary by CodeRabbit

* **New Features**
  * System suffixing for duplicate bank account names and clearer display separating base name from suffix.

* **Improvements**
  * Create/rename now retry to avoid active-name collisions and only enforce uniqueness for active accounts, reducing duplicate errors.

* **Documentation**
  * Added detailed docs explaining unique naming rules, suffixing, and rename behavior.

* **Tests / CI**
  * Added unit tests for name logic and parsing; CI updated to run unit tests.

### CODE REVIEW

**Comment by silviczka** [comment]
Hello @vojtech-cerveny , can you pls check and provide CR.
Thanks!

File: src/domain/bankAccount-domain/ba-repository.ts [hunk 1]
```diff
@@ -99,18 +99,19 @@ export async function createBankAccount({
 
       return bankAccount;
     } catch (error: any) {
-      // Prisma unique constraint error code is P2002
       if (error?.code === "P2002") {
-        throw new Error(
-          `Bank account number "${number}" already exists. Cannot create bank account with duplicate number.`,
-        );
+        const target = error?.meta?.target as string[] | undefined;
```
Comment (vojtech-cerveny) [inline src/domain/bankAccount-domain/ba-repository.ts:103]: nitpick: do we need that? I don't see any benefit of your change here.

File: src/lib/prisma-unique.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/** Returns true when Prisma reports a P2002 unique violation for all listed fields. */
+export function isPrismaUniqueOnFields(error: unknown, fields: string[]): boolean {
```
Comment (vojtech-cerveny) [inline src/lib/prisma-unique.ts:2]: Can you explain this function? :D 



Also the formatting here is not correct, right?

File: src/lib/bank-account-name-display.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/** Matches the app-generated suffix at the end of a name, such as " (01)" or " (123)". */
+export const APP_SUFFIX_REGEX = /\s\((\d{2,})\)$/;
+
+/** Splits a stored account name into its base text and optional app-generated suffix. */
```
Comment (vojtech-cerveny) [inline src/lib/bank-account-name-display.ts:4]: documentation: Some example would be great.

File: src/domain/bankAccount-domain/ba-service.ts [hunk 4]
```diff
@@ -13,23 +14,46 @@ type Pagination = {
...
 
-      const result = await repository.createBankAccount({
-        ...bankAccount,
-        name: finalName,
-        balance: balanceToSet,
-        number: bankAccount.number,
-      });
-      return successResponse("Bank account created successfully", result);
+      let lastError: unknown;
+      for (let attempt = 1; attempt <= MAX_BANK_ACCOUNT_NAME_WRITE_ATTEMPTS; attempt++) {
```
Comment (vojtech-cerveny) [inline src/domain/bankAccount-domain/ba-service.ts:33]: Btw. I still don't get it why we have retry block. What is changed between loops?  We don't use the variable `attempt` anywhere in block, so we are jus trying it multiple times and we expect different results? 🤔 

Can you please explain?

File: src/domain/bankAccount-domain/ba-helpers.ts [hunk 5]
```diff
@@ -33,25 +39,61 @@ export async function enforceZeroBalance(bankAccount: BankAccount) {
...
+ * - Single-digit " (1)" or names like "AKA(1)" are treated as user-entered text, not app suffixes.
+ */
 export async function getUniqueBankAccountName(
   baseName: string,
   userId: string,
   currentAccountId?: string,
 ): Promise<string> {
+  const trimmed = baseName.trim();
+  const appSuffixMatch = trimmed.match(APP_SUFFIX_REGEX);
+  let base = appSuffixMatch ? trimmed.slice(0, -appSuffixMatch[0].length).trimEnd() : trimmed;
```
Comment (vojtech-cerveny) [inline src/domain/bankAccount-domain/ba-helpers.ts:59]: suggestion: maybe some comments in the code would be great if we have slice, trimEnd on one line, or I am just not experienced enough to get it :D

File: prisma/migrations/20260126120000_bankaccount_active_user_name_unique/migration.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+-- Enforce unique (userId, name) only for active accounts so soft-deleted rows do not block reuse.
```
Comment (vojtech-cerveny) [inline prisma/migrations/20260126120000_bankaccount_active_user_name_unique/migration.sql:1]: `20260126120000_bankaccount_active_user_name_unique` - dated January 26, but the PR commits are from March 23. Prisma applies migrations in order by timestamp.
  If any migration was created between January and March, this one will sort before it, which could cause issues. The migration timestamp should be after all existing migrations.

Please rebase the develop and check if this is the problem or not.

File: documentation/bank-account-unique-naming.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# Bank account unique names
```
Comment (vojtech-cerveny) [inline documentation/bank-account-unique-naming.md:1]: Can you add this (shorter version) also into swagger documentation for the endpoint?

**Review by vojtech-cerveny — COMMENTED** [review]
Nice tests, overall I think that it is good, but some code is just too complex to me :D But it can be OK, maybe I just need to know more JS/TS stuff to understand it :)

File: src/domain/bankAccount-domain/ba-repository.ts (hunk 1 above)
Comment (silviczka) [reply src/domain/bankAccount-domain/ba-repository.ts:103]: Fair point. The meta.target check only runs when createBankAccount is called with a custom account number (the first if condition in that function - today that is only seeding, because the UI and the POST /api/v1/bank-account/create endpoint do not send a number).

Before we added the Postgres unique rule on (userId, name) for active accounts only, a P2002 in that situation was almost always “this account number already exists,” so the old duplicate-number message from CZBANK-73 was enough. After that rule was added, the same database insert can fail with P2002 for either a duplicate number or a duplicate active name for that user. The target check avoids showing the duplicate-number message when the problem was really the name, and it avoids turning that into a plain Error so the service can still see a Prisma name violation and retry.

Also, that duplicate-number error message comes from the optional-number work in CZBANK-73 - only seeding uses that today - and we do not really show it in normal UI/API usage, so I understand it can feel like a lot of code for something uncommon.

I agree the benefit is small in real day-to-day use. I am happy to remove the meta.target check if you prefer a simpler repository here - just say what you would like.

File: src/domain/bankAccount-domain/ba-service.ts (hunk 4 above)
Comment (silviczka) [reply src/domain/bankAccount-domain/ba-service.ts:33]: Good question. What changes between iterations is getUniqueBankAccountName() - it is called again inside the loop, so after a (userId, name) P2002 we re-read the DB and can pick a different suffix on the next try. `attempt` is only used for the small backoff `(40 * attempt)`, not to change the naming logic itself.

That said, I agree this is mostly a “just in case” guard. In normal usage the UI disables submit, seeding is sequential, and this would only really matter if two requests for the same user and same name overlap almost at the same moment (for example parallel automated API calls / tests). So looking at our actual usage, I agree this may be unnecessary extra logic. If you want, I can simplify it and just return the error if it happens.

File: src/lib/prisma-unique.ts (hunk 2 above)
Comment (silviczka) [reply src/lib/prisma-unique.ts:2]: This is the helper used by ba-service.ts for the retry and by ba-repository.ts for the target check - I explained both in my other replies above. In short it just checks which fields caused a P2002 so we can tell a number collision from a name collision.

About the formatting - you are right, Prettier flags it. I will run prettier --write on that file. Thanks for catching that!
